send-transaction-service/src/send_transaction_service.rs

```diff
@@ -136,70 +137,88 @@ impl SendTransactionService {
+                //faster then once very 400ms.
+                //if retry_rate_ms is low this stops retries from going faster then once every 400ms.
+                const MAX_DEDUPER_AGE: Duration = Duration::from_millis(400);
+                const MAX_DEDUPER_ITEMS: u32 = 1_000_000;
+                let mut deduper = Deduper::new(MAX_DEDUPER_ITEMS, MAX_DEDUPER_AGE);
+                loop {
+                    deduper.reset();
+                    match receiver
+                        .recv_timeout(Duration::from_millis(1000.min(config.retry_rate_ms)))
+                    {
+                        Err(RecvTimeoutError::Disconnected) => break,
+                        Err(RecvTimeoutError::Timeout) => {}
+                        Ok(transaction_info) => {
+                            inc_new_counter_info!("send_transaction_service-recv-tx", 1);
+                            let addresses = leader_info.as_ref().map(|leader_info| {
+                                leader_info.get_leader_tpus(config.leader_forward_count)
+                            });
+                            let addresses = addresses
+                                .map(|address_list| {
+                                    if address_list.is_empty() {
+                                        vec![&tpu_address]
+                                    } else {
+                                        address_list
+                                    }
+                                })
+                                .unwrap_or_else(|| vec![&tpu_address]);
+                            let mut dedup_fail = 0;
+                            for address in addresses {
+                                dedup_fail += Self::send_transaction(
+                                    &send_socket,
+                                    address,
+                                    &transaction_info.wire_transaction,
+                                    &deduper,
+                                );
+                            }
+                            inc_new_counter_info!(
+                                "send_transaction_service-dupicate",
+                                usize::try_from(dedup_fail).unwrap()
                             );
-                        }
-                        if transactions.len() < MAX_TRANSACTION_QUEUE_SIZE {
-                            inc_new_counter_info!("send_transaction_service-insert-tx", 1);
-                            transactions.insert(transaction_info.signature, transaction_info);
-                        } else {
-                            datapoint_warn!("send_transaction_service-queue-overflow");
+                            if transactions.len() < MAX_TRANSACTION_QUEUE_SIZE {
+                                inc_new_counter_info!("send_transaction_service-insert-tx", 1);
+                                transactions.insert(transaction_info.signature, transaction_info);
```

Comment:
`transactions` is already a `HashMap` (though also vulnerable to sig-malleability) so we already have retry dedupe.  If we're going to do any further dedupe, it should be in front of the initial `send_transaction` call